tools/gradle/plugin/src/main/kotlin/com/zeroc/ice/SliceToolsResourceExtractor.kt

```diff
@@ -0,0 +1,118 @@
+// Copyright (c) ZeroC, Inc.
+
+package com.zeroc.ice
+
+import org.gradle.api.Project
+import java.io.File
+import java.nio.file.Files
+import java.nio.file.StandardCopyOption
+
+/**
+ * Extracts the Slice compiler (slice2java) and Slice files from the plugin resources to the Gradle user home directory.
+ */
+object SliceToolsResourceExtractor {
+    private const val pluginName = "slice-tools"
+
+    /**
+     * Extracts the Slice compiler (slice2java) from the plugin resources to the Gradle user home directory.
+     *
+     * @param project The Gradle project
+     * @return The path to the directory containing the Slice compiler, or null if the extraction failed or the
+     * compiler is not present in the resources.
+     */
+    fun extractSliceCompiler(project: Project): String? {
+        val pluginVersion = getPluginVersion(project)
+
+        val os = detectOS()
+        val arch = detectArch()
+        val toolName = if (os == "windows") "slice2java.exe" else "slice2java"
+        val resourcePath = "/$os-$arch/$toolName"
+
+        val gradleUserDir = File(project.gradle.gradleUserHomeDir, "native/$pluginName-$pluginVersion/bin")
+        val slice2JavaFile = File(gradleUserDir, toolName)
+
+        if (!slice2JavaFile.exists()) {
+            val resourceStream = this::class.java.getResourceAsStream(resourcePath)
+            if (resourceStream != null) {
+                project.logger.lifecycle("Extracting $toolName to ${gradleUserDir.absolutePath}")
+                gradleUserDir.mkdirs()
+                Files.copy(resourceStream, slice2JavaFile.toPath(), StandardCopyOption.REPLACE_EXISTING)
+                slice2JavaFile.setExecutable(true)
+            } else {
+                return null
+            }
+        }
+
+        return gradleUserDir.absolutePath
+    }
+
+    /**
+     * Extracts the Slice files from the plugin resources to the Gradle user home directory.
+     *
+     * @param project The Gradle project
+     * @return The path to the directory containing the Slice files, or null if the extraction failed or the
+     * Slice files are not present in the resources.
+     */
+    fun extractSliceFiles(project: Project): String? {
+        val pluginVersion = getPluginVersion(project)
+
+        val sliceSourcePath = "/slice/"
+        val sliceTargetDir = File(project.gradle.gradleUserHomeDir, "native/$pluginName-$pluginVersion/slice")
+
+        if (!sliceTargetDir.exists()) {
+            val resourceURL = this::class.java.getResource(sliceSourcePath)
```

Comment:
This is some sort of special resource that we can't just use without "copying"?. I.e. returning `File(resourceURL.toURI())`